modules/uploads/scripts/delete.ts

```diff
@@ -0,0 +1,86 @@
+import { RuntimeError, ScriptContext } from "../_gen/scripts/delete.ts";
+import { getKey } from "../utils/types.ts";
+import { deleteKeys } from "../utils/bucket.ts";
+import { getS3EnvConfig } from "../utils/env.ts";
+
+export interface Request {
+	uploadId: string;
+}
+
+export interface Response {
+	bytesDeleted: string;
+}
+
+export async function run(
+	ctx: ScriptContext,
+	req: Request,
+): Promise<Response> {
+	const s3 = getS3EnvConfig(ctx.userConfig.s3);
+	if (!s3) throw new RuntimeError("s3_not_configured");
+
+	const bytesDeleted = await ctx.db.$transaction(async (db) => {
+		const upload = await db.upload.findFirst({
+			where: {
+				id: req.uploadId,
+				completedAt: { not: null },
+				deletedAt: null,
+			},
+			select: {
+				id: true,
+				userId: true,
+				bucket: true,
+				contentLength: true,
+				files: true,
+				createdAt: true,
+				updatedAt: true,
+				completedAt: true,
+			},
+		});
+		if (!upload) {
+			throw new RuntimeError(
+				"upload_not_found",
+				{
+					meta: {
+						modified: false,
+						uploadId: req.uploadId,
+					},
+				},
+			);
+		}
+
+		const filesToDelete = upload.files.map((file) =>
+			getKey(file.uploadId, file.path)
+		);
+		const deleteResults = await deleteKeys(s3, filesToDelete);
+
+		const failures = upload.files
+			.map((file, i) => [file, deleteResults[i]] as const)
```

Comment:
why do we need as const?

Reply:
Typescript type wonderfulness:
An example:
```ts
let a = [0, "hello"]; // a is (string | number)[]
let b = [0, "hello"] as const; // b is readonly [number, string]
let c = [...([0, "hello"] as const)]; // c is [number, string]
```

Typescript doesn't do lookahead type inference— by default, an array  literal becomes an array of the union of the types of its members.
However, you can assert to typescript that you want it as a readonly tuple instead of as an array by adding `as const` to it.

That said, to remove that wackiness, maybe I should switch it to a for loop for ease of understanding.

Reply:
oh til another weird typescript weirdity